### PR TITLE
Fix for #384 - And some other changes to controller as filter

### PR DESCRIPTION
**Pull request by rynowak**
This is a major change to how we handle the scenario where a controller is
a filter. We want to change the lifetime of the controller object, by
scoping it around action filters and result filters. This means that a
controller class can only implement action filters and result filters.

To implement #384 - we're creating a delegating filter class
'ControllerFilter' which will forward calls to the implementation of the
controller. This is discovered in the controller model and added to the
filter collection. This filter is removable as an opt-out of this feature.

The ControllerFilter only implements action filter and result filter, so
the new restriction about filter types on Controller is in place. A future
change will move the instantiation of the controller to after resource
filters.

### CODE REVIEW

**Comment by rynowak** [comment]
Ping

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs [hunk 1]
```diff
@@ -141,6 +142,18 @@ protected virtual ControllerModel CreateControllerModel([NotNull] TypeInfo typeI
...
+            // Controllers can implement action filter and result filter interfaces. We add
+            // a special delegating filter implementation to the pipeline to handle it.
+            //
+            // This is needed because filters are instantiated before the controller.
+            if (typeof(IAsyncActionFilter).GetTypeInfo().IsAssignableFrom(typeInfo) ||
+                typeof(IActionFilter).GetTypeInfo().IsAssignableFrom(typeInfo) ||
+                typeof(IAsyncResultFilter).GetTypeInfo().IsAssignableFrom(typeInfo) ||
+                typeof(IResultFilter).GetTypeInfo().IsAssignableFrom(typeInfo))
+            {
+                controllerModel.Filters.Add(new ControllerFilter());
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: Split

File: src/Microsoft.AspNet.Mvc.Core/Filters/ControllerFilter.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    /// interfaces.
+    /// </summary>
+    public class ControllerFilter : IAsyncActionFilter, IAsyncResultFilter, IOrderedFilter
+    {
+        // Controller-filter methods run farthest from the action by default.
+        /// <inheritdoc />
+        public int Order { get; set; } = int.MinValue;
+
+        /// <inheritdoc />
+        public async Task OnActionExecutionAsync([NotNull] ActionExecutingContext context, [NotNull] ActionExecutionDelegate next)
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ControllerFilter.cs]: linelength

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultControllerModelBuilderTest.cs [hunk 3]
```diff
@@ -220,4 +271,27 @@ public class NoSuffix : Mvc.Controller
     public class PocoController
     {
     }
+
+    [Produces("application/json")]
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultControllerModelBuilderTest.cs]: Controller with a filter type that is not supported does not show up (and consider if we should log/trace)

**Comment by yishaigalatzer** [comment]
:shipit: when comments are addressed.

**Pull request closed without merging**